bigtable/admin.go

```diff
@@ -376,79 +439,124 @@ func (ac *AdminClient) CreateColumnFamilyWithConfig(ctx context.Context, table,
-}
-
-// UpdateTableWithChangeStream updates a table to with the given table ID and change stream config.
-func (ac *AdminClient) UpdateTableWithChangeStream(ctx context.Context, tableID string, changeStreamRetention ChangeStreamRetention) error {
-	return ac.updateTableWithConf(ctx, &UpdateTableConf{tableID, None, changeStreamRetention})
-}
-
-// UpdateTableWithDeletionProtection updates a table with the given table ID and deletion protection parameter.
-func (ac *AdminClient) UpdateTableWithDeletionProtection(ctx context.Context, tableID string, deletionProtection DeletionProtection) error {
-	return ac.updateTableWithConf(ctx, &UpdateTableConf{tableID, deletionProtection, nil})
-}
+const (
+	deletionProtectionFieldMask    = "deletion_protection"
+	changeStreamRetentionFieldMask = "change_stream_config"
+	automatedBackupPolicyFieldMask = "automated_backup_policy"
+)
 
-// updateTableWithConf updates a table in the instance from the given configuration.
-// only deletion protection can be updated at this period.
-// table ID is required.
-func (ac *AdminClient) updateTableWithConf(ctx context.Context, conf *UpdateTableConf) error {
-	if conf.tableID == "" {
-		return errors.New("TableID is required")
+func newUpdateTableRequestProto(tableID string, prefix string) (*btapb.UpdateTableRequest, error) {
+	if tableID == "" {
+		return nil, errors.New("TableID is required")
 	}
-
-	ctx = mergeOutgoingMetadata(ctx, ac.md)
-
 	updateMask := &field_mask.FieldMask{
 		Paths: []string{},
 	}
-	prefix := ac.instancePrefix()
 	req := &btapb.UpdateTableRequest{
 		Table: &btapb.Table{
-			Name: prefix + "/tables/" + conf.tableID,
+			Name: prefix + "/tables/" + tableID,
 		},
 		UpdateMask: updateMask,
 	}
+	return req, nil
+}
 
-	if conf.deletionProtection != None {
-		updateMask.Paths = append(updateMask.Paths, "deletion_protection")
-		req.Table.DeletionProtection = conf.deletionProtection != Unprotected
+// updateTable updates a table in the instance, with the given UpdateTableRequest
+func (ac *AdminClient) updateTable(ctx context.Context, tableID string, updateTableRequestLogic func(utr *btapb.UpdateTableRequest) error) error {
+	updateTableRequest, err := newUpdateTableRequestProto(tableID, ac.instancePrefix())
+	if err != nil {
+		return fmt.Errorf("error when updating table: %w", err)
 	}
-
-	if conf.changeStreamRetention != nil {
-		if conf.changeStreamRetention.(time.Duration) == time.Duration(0) {
-			updateMask.Paths = append(updateMask.Paths, "change_stream_config")
-		} else {
-			updateMask.Paths = append(updateMask.Paths, "change_stream_config.retention_period")
-			req.Table.ChangeStreamConfig = &btapb.ChangeStreamConfig{}
-			req.Table.ChangeStreamConfig.RetentionPeriod = durationpb.New(conf.changeStreamRetention.(time.Duration))
-		}
+	err = updateTableRequestLogic(updateTableRequest)
+	if err != nil {
+		return fmt.Errorf("error when updating table: %w", err)
 	}
+	return ac.updateTableAndWait(ctx, updateTableRequest)
```

Comment:
I dont think inversion of control is necessary here...I think it would be easier to do this:

```go
func (ac *AdminClient) UpdateTableWithDeletionProtection(ctx context.Context, tableID string, deletionProtection DeletionProtection) error {
        req, err := ac.newUpdateTableRequestProto(tableID)
        if (err != nil) {
           return err
        }
        req.UpdateMask.Paths = [] str {deletionProtectionFieldMask}
        req.Table.DeletionProtection = deletionProtection != Unprotected
        return updateTableAndWait(ctx, req)
}
```

Reply:
Thanks, fixed